Generation: stop at `eos` for assisted decoding

zucchini-nlp opened the pull request:
# What does this PR do?

Addresses the issue in [this comment](https://github.com/huggingface/transformers/pull/27775#issuecomment-2151908426). 

The problem was that assisted generation is flaky when it comes to EOS tokens. If the assistant generates target model's `eos` token in the middle of candidate sequence and it get accepted by the target, we don't stop generating because the criteria checks only the last token. This happens mostly for Prompt-Lookup decoding because it's simple copy-paste from history.

One possible solution is to check `eos` in all of the newly generated input ids, another is to crop candidates if there's `eos`.  I went the second way.

zucchini-nlp committed:
fix

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31301). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
Merge branch 'huggingface:main' into promp_lookup_decoding

Ednaordinary commented:
Latest release doesn't have a fix for this, it's still an issue

zucchini-nlp committed:
Merge branch 'huggingface:main' into promp_lookup_decoding

zucchini-nlp committed:
Merge branch 'huggingface:main' into promp_lookup_decoding

gante reviewed (commented):
Thank you for taking this issue!

the logic LGTM, I'd like to request two changes for future-proofing:
1. This issue is unique to prompt lookup (the original assisted generation with an assistant model stops if the assistant generates an eos token). Can we move the logic to the corresponding candidate generator?
2. Add a test 👼

zucchini-nlp committed:
move changes to prompt lookup

zucchini-nlp committed:
add test

zucchini-nlp committed:
set eos in assistant model

zucchini-nlp committed:
style

zucchini-nlp commented:
@gante okay, moved the logic into PLD. The main concern was that assistant model can have it's eos token to be different from target model, so I added a line in AssistedGenerator to set eos token. 

Added a test for PLD, but not for assistant model because we can't control what the assistant model generates. Also tested with the case from the linked issue, is working on my end

zucchini-nlp committed:
fix flakiness

zucchini-nlp committed:
Merge branch 'huggingface:main' into promp_lookup_decoding

gante reviewed (approved):
Thank you for iterating and fixing the issue 💛

Ednaordinary commented:
Please note this will error out if the eos_token_id is a list (like in some llama3 configurations)
![image](https://github.com/user-attachments/assets/c88787ab-5a75-4f9e-8620-73abd59f28fd)

zucchini-nlp commented:
@Ednaordinary are you using the `generate()` method? I don't see and error using current branch, because internally we convert all special tokens to tensors and `self.eos_id` cannot be a list.

I guess you're using `main` and applied current changes above, in that case it will fail because I didnt yet rebase. There were some changes for special tokens lately afaik. I will make sure it works before merging

Ednaordinary commented:
In some configurations running llama, the eos_token_id is set to a list.
![image](https://github.com/user-attachments/assets/0f04508f-34f8-4e84-a6d3-827173eec32d)
this works fine for sampling until prompt lookup decoding is added. It's a simple fix, though, adding [0] to the end of tokenizer.encode("<|eot_id|>"). Works fine after applying this PR to current main
I mainly raise this concern because llama 3.1 has three stop tokens:
![image](https://github.com/user-attachments/assets/2a310d49-de8a-4643-aff4-69da6188e208)

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into promp_lookup_decoding

zucchini-nlp committed:
changes for new `main`

zucchini-nlp commented:
@Ednaordinary I see, you're using llama-3.1 that forces you to use the latest version. As I said, the problem was that current branch is behind `main` for months. Now I rebased and everything should be fine

amyeroberts commented on tests/generation/test_utils.py:
```diff
@@ -1372,6 +1373,26 @@ def test_assisted_decoding_sample(self):
+        eos_token_id = torch.tensor([0], device=torch_device)
+
+        # init cand geenerator with max_matching_ngram_size=1 to match per-token
+        cand_generator = PromptLookupCandidateGenerator(
```
We can use the full name here - we're not paying per character! :D 

```suggestion
        candidate_generator = PromptLookupCandidateGenerator(
```

amyeroberts commented on tests/generation/test_utils.py:
```diff
@@ -1372,6 +1373,26 @@ def test_assisted_decoding_sample(self):
+
+        # init cand geenerator with max_matching_ngram_size=1 to match per-token
+        cand_generator = PromptLookupCandidateGenerator(
+            eos_token_id=eos_token_id, num_output_tokens=4, max_matching_ngram_size=1
```
Just to make sure I've understood correctly - the value of `num_output_tokens` is independent of the position we injected the eos token id, right (`input_ids[:, 4] = 0`) ?

zucchini-nlp replied:
Yes, in any case we should not be proposing new candidates, which means we return the input as is. I can even make a check that we got same values as input ids

amyeroberts commented on tests/generation/test_utils.py:
```diff
@@ -1372,6 +1373,26 @@ def test_assisted_decoding_sample(self):
 
             self._check_outputs(output_assisted, input_ids, model.config, use_cache=True)
 
+    def test_prompt_lookup_decoding_stops_at_eos(self):
```
to make this test clearer and more robust, could we first get the shape of the output_prompt_lookup without an eos token id injected, and then show the difference once we add it?

amyeroberts replied:
This indicates to me we're not testing or not able to test the right thing: the tests is called `test_prompt_lookup_decoding_stops_at_eos`, but if eos being present or not doesn't change anything I don't see how we're ensuring this?

zucchini-nlp replied:
I think we are testing that the buggy behavior is not there anymore, so its output is 10 when we trigger setting when bug was found. That is actually all we can test :(

amyeroberts replied:
OK :(

amyeroberts reviewed (commented):
Thanks for handling this! 

Just a few small comments and questions

zucchini-nlp commented on src/transformers/generation/candidate_generator.py:
```diff
@@ -319,6 +324,15 @@ def get_candidates(self, input_ids: torch.LongTensor) -> Tuple[torch.LongTensor,
+
+                    # remove remaining candidate ids if an "eos" token is found, otherwise the target model may
+                    # accept eos and the rest as valid, thus not stopping generation after "eos"
+                    # NOTE: below code is written based on the fact that assisted decoding supports only bs=1
```
This will work in all cases, if the input contains and eos or doesn't. It not always the case that an eos is chosen as candidate, that depends on the overlap between prompt and the generated part

amyeroberts replied:
Ah, sorry, I meant re batch size 1, can we assume this will always hole?

zucchini-nlp replied:
Yes, the current impl doesn;t support more than 1. If anyone adds support, this and other parts of postprocessing will have to be modified

zucchini-nlp committed:
Update tests/generation/test_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update tests/generation/test_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp commented on tests/generation/test_utils.py:
```diff
@@ -1372,6 +1373,26 @@ def test_assisted_decoding_sample(self):
+
+        # init cand geenerator with max_matching_ngram_size=1 to match per-token
+        cand_generator = PromptLookupCandidateGenerator(
+            eos_token_id=eos_token_id, num_output_tokens=4, max_matching_ngram_size=1
```
To make the test less confusing and check unigram matches, otherwise we have to inject ngram that contains an 'eos' so that the last ngram of input is same as the injected one. Too many dependencies and prone to error when crafting

zucchini-nlp committed:
add comment to explain

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into promp_lookup_decoding
